File: apps/api/schemas/lineup.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from models.enums import GameweekStatus, Position


class LineupPlayerResponse(BaseModel):
    """One player in the gameweek lineup."""
    player_id: int
    name: str
    display_name: Optional[str]
    position: Position
    team_name: str
    team_short_name: str
    current_price: float
    form_score: float
    is_available: bool
    is_captain: bool
    is_vice_captain: bool
    points_scored: Optional[int]  # filled by scoring job; null until then

    model_config = {"from_attributes": False}


class LineupResponse(BaseModel):
    """
    Full lineup response — returned by GET and PUT /lineups/current.

    is_editable: convenience flag the frontend uses to enable/disable the picker.
    It is False when gameweek_status is LOCKED, ACTIVE, SCORING, or FINISHED,
    or when the lineup's own is_locked flag is True.
    """
    id: int
    gameweek_id: int
    gameweek_number: int
    gameweek_name: str
    gameweek_deadline: datetime
    gameweek_status: GameweekStatus
    is_locked: bool
    is_editable: bool
    captain_player_id: Optional[int]
    vice_captain_player_id: Optional[int]
    players: list[LineupPlayerResponse]
    points_scored: Optional[int]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": False}


class CaptainUpdateRequest(BaseModel):
    """Payload for PUT /lineups/current — set captain and vice-captain."""
    captain_player_id: int = Field(description="Player ID to assign as captain (2× points).")
    vice_captain_player_id: int = Field(
        description="Player ID to assign as vice-captain. "
                    "Active only if captain did not appear (0 minutes played)."
    )


class LineupValidationErrorResponse(BaseModel):
    """Returned when captain/vc validation fails (HTTP 400)."""
    detail: str = "Lineup validation failed."
    errors: list[str]

File: apps/web/src/lib/api/lineup.ts
import { apiFetch, ApiError } from './client'

export type GameweekStatus = 'UPCOMING' | 'LOCKED' | 'ACTIVE' | 'SCORING' | 'FINISHED'
export type Position = 'GK' | 'DEF' | 'MID' | 'FWD'

export interface LineupPlayer {
  player_id: number
  name: string
  display_name: string | null
  position: Position
  team_name: string
  team_short_name: string
  current_price: number
  form_score: number
  is_available: boolean
  is_captain: boolean
  is_vice_captain: boolean
  points_scored: number | null
}

export interface LineupResponse {
  id: number
  gameweek_id: number
  gameweek_number: number
  gameweek_name: string
  gameweek_deadline: string
  gameweek_status: GameweekStatus
  is_locked: boolean
  is_editable: boolean
  captain_player_id: number | null
  vice_captain_player_id: number | null
  players: LineupPlayer[]
  points_scored: number | null
  created_at: string
  updated_at: string
}

export { ApiError }

export async function fetchCurrentLineup(): Promise<LineupResponse | null> {
  try {
    return await apiFetch<LineupResponse>('/api/v1/lineups/current')
  } catch (err) {
    if (err instanceof ApiError && (err.status === 404 || err.status === 503)) return null
    throw err
  }
}

export async function updateCaptain(
  captainPlayerId: number,
  vcPlayerId: number,
): Promise<LineupResponse> {
  return apiFetch<LineupResponse>('/api/v1/lineups/current', {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      captain_player_id: captainPlayerId,
      vice_captain_player_id: vcPlayerId,
    }),
  })
}
